src/vdaf/prio3.rs

```diff
@@ -779,23 +779,105 @@ impl HelperShare {
 #[cfg(test)]
 mod tests {
     use super::*;
+    use crate::vdaf::{run_vdaf, run_vdaf_prepare};
 
     #[test]
     fn test_prio3_count64() {
-        let prio3 = Prio3Count64::new(Suite::Blake3, 23).unwrap();
+        let prio3 = Prio3Count64::new(Suite::Blake3, 2).unwrap();
+
+        assert_eq!(
+            run_vdaf(&prio3, &(), [1, 0, 0, 1, 1]).unwrap(),
+            Prio3Result(3)
+        );
+
         let (_, verify_params) = prio3.setup().unwrap();
         let nonce = b"This is a good nonce.";
+
+        let input_shares = prio3.shard(&(), &0).unwrap();
+        run_vdaf_prepare(&prio3, &verify_params, &(), nonce, input_shares).unwrap();
+
         let input_shares = prio3.shard(&(), &1).unwrap();
-        eval_vdaf(
-            &prio3,
-            &input_shares,
-            &verify_params,
-            nonce,
-            Some(Prio3Result(1)),
-        )
-        .unwrap();
-
-        // TODO(cjpatton) Add failure test cases.
+        run_vdaf_prepare(&prio3, &verify_params, &(), nonce, input_shares).unwrap();
+    }
+
+    #[test]
+    fn test_prio3_sum64() {
+        let prio3 = Prio3Sum64::new(Suite::Blake3, 3, 16).unwrap();
+
+        assert_eq!(
+            run_vdaf(&prio3, &(), [0, (1 << 16) - 1, 0, 1, 1]).unwrap(),
+            Prio3Result((1 << 16) + 1)
+        );
+
+        let (_, verify_params) = prio3.setup().unwrap();
+        let nonce = b"This is a good nonce.";
+
+        // TODO Check that the expected error is returned (requires `VdafError` to implement `Eq`).
+        let mut input_shares = prio3.shard(&(), &1).unwrap();
+        input_shares[0].blind.as_mut_slice()[0] ^= 255;
+        run_vdaf_prepare(&prio3, &verify_params, &(), nonce, input_shares).unwrap_err();
+
+        let mut input_shares = prio3.shard(&(), &1).unwrap();
+        input_shares[0].joint_rand_seed_hint.as_mut_slice()[0] ^= 255;
+        run_vdaf_prepare(&prio3, &verify_params, &(), nonce, input_shares).unwrap_err();
+
+        let mut input_shares = prio3.shard(&(), &1).unwrap();
+        match input_shares[0].input_share {
+            Share::Leader(ref mut data) => {
+                data[0] += Field96::one();
+            }
+            _ => panic!("expected leader share"),
+        }
```

Comment:
`assert_matches!` would work well here. You can even get a scope where you can work with bindings of the destructured pattern, as seen [here](https://github.com/abetterinternet/ppm-prototype/blob/a0ed255336762ac1ef1ea08b459cb725996851d5/tests/tests.rs#L201).